Introduce Cloud docker variant

pugnascotia opened the pull request:
Closes #74795.

Introduce two Docker image variants for Cloud. The first bundles (actually installs) the S3, Azure and GCS repository plugins. The second bundles all official plugins, but only installs the repository plugins.

Both images also bundle Filebeat and Metricbeat.

The testing utils have been refactored to introduce a `docker` sub-package. This allows the static `Docker.containerId` to be shared without needing all the code in one big class. The code for checking file ownership / permissions has also been refactored to a more Hamcrest style, using a custom Docker file matcher.

pugnascotia committed:
Introduce Cloud docker variant

pugnascotia committed:
Preinstall repository plugins in Cloud

pugnascotia committed:
Don't bother with DockerHub for plugins

pugnascotia committed:
Copy in beats

pugnascotia committed:
Tweaks

pugnascotia committed:
Tweak

pugnascotia committed:
Fix packaging tests

pugnascotia committed:
More testing

pugnascotia committed:
More tweaks

pugnascotia committed:
More tweaks

pugnascotia committed:
Upgrade curl

pugnascotia committed:
Revert some changes

elasticmachine commented:
Pinging @elastic/es-delivery (Team:Delivery)

mark-vieira commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -257,6 +259,24 @@ RUN sed -i -e 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' bin/elas
+<% if (docker_base == "cloud") { %>
+# Preinstall common plugins
+COPY repository-s3-${version}.zip repository-gcs-${version}.zip repository-azure-${version}.zip /opt/
+RUN bin/elasticsearch-plugin install --batch \\
```
Should we delete these files after install?

pugnascotia replied:
This happens in the `builder` image, which we throw away.

mark-vieira replied:
Ah, right.

pugnascotia committed:
Revert change

pugnascotia committed:
Bundle complete beats distro in Cloud image

Also reorganize Docker testing classes into their own sub-package, and
sub-class the `FileMatcher` utility instead of using a bespoke assertion
method.

pugnascotia committed:
Adjust entrypoint and cmd for Cloud image

pugnascotia committed:
Sort imports

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 74795-cloud-docker-image

alpar-t commented on distribution/docker/build.gradle:
```diff
@@ -40,40 +48,43 @@ configurations {
+  repositoryS3Plugin project(path: ':plugins:repository-s3', configuration: 'zip')
+  repositoryGcsPlugin project(path: ':plugins:repository-gcs', configuration: 'zip')
+  repositoryAzurePlugin project(path: ':plugins:repository-azure', configuration: 'zip')
+  filebeat "beats:filebeat:${VersionProperties.elasticsearch}:${beatsArch}@tar.gz"
```
Does this work itself out to a version with a build ID in case of release manager builds? I think we need that to guarantee consistency in the unified release process and make sure that we are picking up the beats from the same build. Also means that ES needs to be built after beats as part of that process. 

Also, someone on the beats team might want to customize this, e.g. check out some ES tag to get a stable/known to be good ES and build the docker images with a custom working version of beats. That's a use case for PRs against beats, but I think it's this code that needs to enable it somehow. Just a taught, it can be done in a follow up PR too.

pugnascotia replied:
The corresponding release-manager changes fetch the beats artifacts using the build ID, so everything should match up.

alpar-t commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -257,6 +259,21 @@ RUN sed -i -e 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' bin/elas
 
+<% if (docker_base == "cloud") { %>
+# Preinstall common plugins
+COPY repository-s3-${version}.zip repository-gcs-${version}.zip repository-azure-${version}.zip /tmp/
```
This is a question for @mieciu ho's probably on top of this already. Will this work with the existing startup script in cloud ? Once we start using these images, they need to work with already released versions of ECE so while we can and should  make improvements to this script on the cloud side, we have to be backwards compatible too. I'm thinking we might also need a way to tell cloud here that these are already installed.

mieciu replied:
Indeed, that's something that's not going to work for us, at least for `7.x`.

pugnascotia replied:
Can you advise on what would work?

pugnascotia replied:
Perhaps `allocator.sh` could be patched to never add or remove these plugins?

alpar-t replied:
I think that's the way to go, I pinged @mieciu because I think that can be handled on the cloud side. We add this script as part of the stack-pack so we can make changes to it that will take affect on older allocators too.

pugnascotia replied:
If it's tricky to modify `allocator.sh` for some reason, we could add logic to `plugin-wrapper.sh` here.

pugnascotia committed:
Change beats repo if property is set

alpar-t commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -257,6 +259,21 @@ RUN sed -i -e 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' bin/elas
     find . -type f -exec chmod o+r {} +
 
+<% if (docker_base == "cloud") { %>
+# Preinstall common plugins
```
Has there been a discussion around other plugins that are now part of the cloud image and customers can install ? 
AFAIK some of the popular examples are the `analysis-*` plugins.  We have options either way: bundle them in the image, but don't install them by default, or have cloud install them over the download service ( and document that it doesn't work for air gaped installation, these would install it like they install a custom plugin ). 7.x and ECE 2.x are think sensible to this approach we would need to seek PM approval as it would I think be a breaking change for such environments I'm not sure it's ok to make or not.

pugnascotia replied:
I wasn't involved in the discussions to date, I was going by #74795. Once we have something work end-to-end for these plugins, from the ES perspective it ought to be straightforward to add more.

mark-vieira replied:
We discussed this an concluded that for non-essential plugins (ones that aren't _always_ installed for any given cloud provider) they would simply be downloaded as with an on-prem install.

pugnascotia replied:
After further discussion, this PR now builds a `cloud` image and a `cloud-ess` image. The former image installs only the 3 repository plugins, and a plugin tool wrapper that checks in an archive directory for other plugins before reaching out to the internet. The latter image includes all the other plugins in the archive directory, in a "batteries-included" manner.

pugnascotia committed:
Fixes

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 74795-cloud-docker-image

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 74795-cloud-docker-image

pugnascotia committed:
First go at an all-in-one Cloud image

pugnascotia committed:
Add plugin utility wrapper.sh

pugnascotia committed:
Include plugin wrapper in Cloud image, not ESS

pugnascotia committed:
Test coverage for ESS image

pugnascotia committed:
Checkstyle

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 74795-cloud-docker-image

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -16,20 +15,39 @@ apply plugin: 'elasticsearch.test.fixtures'
-// Define a repository that allows Gradle to fetch a resource from GitHub. This
-// is only used to fetch the `tini` binary, when building the Iron Bank docker image
-// for testing purposes.
+String buildId = System.getProperty('build.id')
```
We should use `providers.systemProperty()` here. Do a search to see examples of other usages in our build.

pugnascotia replied:
Changed to use a provider.

mark-vieira commented on distribution/docker/build.gradle:
```diff
@@ -266,31 +303,39 @@ void addUnpackDockerContextTask(Architecture architecture, DockerBase base) {
+  String image = "elasticsearch${base.suffix}"
+
+  String namespace = 'elasticsearch'
+  if (base == DockerBase.CLOUD || base == DockerBase.CLOUD_ESS) {
```
Is it strictly necessary that we do this here? Isn't this done by release manager?

pugnascotia replied:
As far as I can tell, the release-manager build file just contains what image to pick up, it doesn't do the creation or tagging.

mark-vieira commented on build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/DockerBase.java:
```diff
@@ -18,7 +18,13 @@
+    CLOUD("centos:8", "-cloud"),
+
+    // Based on CLOUD above, with more extras
+    CLOUD_ESS("centos:8", "-cloud-ess");
```
Do we actually use this base image bit anywhere. I feel like we end up overriding this in `docker/build.gradle` anyhow.

pugnascotia replied:
We reference `base.image` in a number of places in the docker `build.gradle`.

mark-vieira replied:
Right, but we end up overwriting it [here](https://github.com/elastic/elasticsearch/pull/74980/files#diff-39ead019dc289b9b901bfa7df8efc01b1f040df71c88919b5fc7a6286b0d6597R382). Anyway to avoid that?

pugnascotia replied:
Oh sorry, you mean in this specific case. Yes, we might as well not set it and comment why.

mark-vieira commented on plugins/build.gradle:
```diff
@@ -22,4 +27,9 @@ configure(subprojects.findAll { it.parent.path == project.path }) {
   }
+
+  String configurationName = project.name.startsWith('repository-') ? 'repositoryPlugins' : 'nonRepositoryPlugins'
+  parent.dependencies { DependencyHandler d ->
```
This doesn't seem right. We're adding dependencies to these configurations here _and_ resolving them from another project, which I'm pretty sure is a no-no. We should be instead adding the plugins as _artifacts_ to these configurations for external consumptions. @breskeby that sound right?

pugnascotia replied:
I swapped this our for artifacts - any better?

mark-vieira replied:
Yup, that should do the trick.

mark-vieira reviewed (changes requested):
Left some minor feedback.

pugnascotia committed:
Address review feedback

pugnascotia committed:
Fix

elasticmachine committed:
Merge branch 'master' into 74795-cloud-docker-image

pugnascotia committed:
Don't set a base image for CLOUD_ESS

pugnascotia commented on distribution/docker/src/docker/Dockerfile:
```diff
@@ -259,6 +261,27 @@ RUN sed -i -e 's/ES_DISTRIBUTION_TYPE=tar/ES_DISTRIBUTION_TYPE=docker/' bin/elas
+COPY filebeat-${version}.tar.gz metricbeat-${version}.tar.gz /tmp/
+RUN mkdir -p /opt/filebeat /opt/metricbeat && \\
+    tar xf /tmp/filebeat-${version}.tar.gz -C /opt/filebeat --strip-components=1 && \\
+    tar xf /tmp/metricbeat-${version}.tar.gz -C /opt/metricbeat --strip-components=1
```
@alpar-t @mieciu I'd still like to know what Cloud actually needs from the Beats distributions. I assume the entire distro isn't required. This isn't essential now, but anything we can trim from the image will help to keep the size down.

pugnascotia committed:
Tweaks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 74795-cloud-docker-image

mark-vieira reviewed (approved):
:+1:
